subsys/net/lib/lwm2m/lwm2m_rd_client.c

```diff
@@ -137,27 +126,28 @@ static void set_sm_state(int index, u8_t state)
 	clients[index].engine_state = state;
 }
 
+static bool sm_is_registered(int index)
+{
+	return (clients[index].engine_state >= ENGINE_REGISTRATION_DONE &&
+		clients[index].engine_state <= ENGINE_DEREGISTER_FAILED);
+}
+
 static u8_t get_sm_state(int index)
 {
 	/* TODO: add locking? */
 	return clients[index].engine_state;
 }
 
-static int find_clients_index(const struct sockaddr *addr,
-				  bool check_bs_server)
+static int find_clients_index(const struct sockaddr *addr)
 {
 	int index = -1, i;
+	struct net_app_ctx *app_ctx;
 
 	for (i = 0; i < client_count; i++) {
-		if (check_bs_server) {
-			if (memcmp(addr, &clients[i].bs_server,
-				   sizeof(addr)) == 0) {
-				index = i;
-				break;
-			}
-		} else {
-			if (memcmp(addr, &clients[i].reg_server,
-				   sizeof(addr)) == 0) {
+		app_ctx = &clients[i].ctx->net_app_ctx;
+		if (clients[i].ctx) {
+			if (memcmp(addr, &app_ctx->default_ctx->remote,
+				   sizeof(*addr)) == 0) {
```

Comment:
I just realize that in IPv6 case, the sockaddr in the app_ctx doesn't have the "sin6_scope_id" setup while "addr" does.
There's no match even if the packet is coming from the same peer.

Reply:
Yep, I am going to update this patch to use net_ipv4_addr_cmp() / net_ipv6_addr_cmp().